### ISSUE TITLE
[YSQL] Unique constraint not being enforced correctly

### ISSUE THREAD

**Issue by kmuthukk** [YugaByte/yugabyte-db#1434]
Test Case:

```
# Dependencies:
# On CentOS you can install psycopg2 thus:
#
#  sudo yum install postgresql-libs
#  sudo yum install python-psycopg2

import psycopg2;
from multiprocessing.dummy import Pool as ThreadPool


# Two threads inserting unique user-ids, but overlapping user names.
# We have a UNIQUE index on user name. So half of the inserts should
# fail with UniqueConstraintViolation error.
#
# Thread 0 attempts to insert tuples like:
#    user-0-0, name-0
#    user-0-1, name-1
#    user-0-2, name-2
#
# Thread 1 attempts to insert tuples like:
#    user-1-0, name-0
#    user-1-1, name-1
#    user-1-2, name-2
#
# Each thread attempts to insert 1000 rows (so total 2000 attempts).
# But half of those should fail with Unique Contraint Violation
# and we should have 1000 rows finally.
#
# However, we seem to end up sometimes with more rows (like 1001 or
# 1003 etc.)
#
num_users=1000

def create_table():
  conn = psycopg2.connect("host=localhost dbname=postgres user=postgres port=5433")
  conn.set_session(autocommit=True)
  cur = conn.cursor()
  cur.execute("""DROP TABLE IF EXISTS users""");
  cur.execute("""CREATE TABLE IF NOT EXISTS users(
                      id     text,
                      ename  text,
                      age    int,
                      PRIMARY KEY(id))
              """)
  print("Created users table")
  print("====================")
  cur.execute("""
          CREATE UNIQUE INDEX IF NOT EXISTS name_idx ON users(ename)
    """)
  print("Created name_idx on table")



def load_data_slave(thread_num):
  thread_id = str(thread_num)
  conn = psycopg2.connect("host=localhost dbname=postgres user=postgres port=5433")
  conn.set_session(autocommit=True)
  cur = conn.cursor()

  print("Thread-" + thread_id + ": Inserting %d rows..." % (num_users))
  # "name" column is indexed, and may have duplicates across threads.
  num_inserts = 0
  num_uniq_violations = 0
  try:
   for idx in range(num_users):
     try:
       cur.execute("""INSERT INTO users (id, ename, age) VALUES (%s, %s, %s)""",
                   ("user-"+thread_id+"-"+str(idx),
                    "name--"+str(idx),
                     20 + (idx % 50)))
       num_inserts += 1
     except psycopg2.errors.UniqueViolation:
       print("Expected unique constraint violation; thread: " + str(thread_num))
       num_uniq_violations += 1
  except Exception as e:
    print("Unexpected exception: " + str(e))

  print("Thread-" + thread_id + ": Inserted %d rows" %  (num_inserts))
  print("Thread-" + thread_id + ": Uniq Violations: %d " %  (num_uniq_violations))

def load_data():
  pool = ThreadPool(2)
  results = pool.map(load_data_slave, range(2))

# Main
create_table()
load_data()
```

Output:

```
01:15 $ python ~/notes/ysql_sec_index.py
Created users table
====================
Created name_idx on table
Thread-1: Inserting 1000 rows...
Thread-0: Inserting 1000 rows...
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 1
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Expected unique constraint violation; thread: 0
Unexpected exception: Query error: Restart read required at: { read: { physical: 1558746958970781 } local_limit: { physical: 1558746958971794 } global_limit: <min> in_txn_limit: <max> serial_no: 0 }

Thread-0: Inserted 18 rows
Thread-0: Uniq Violations: 10
Thread-1: Inserted 984 rows
Thread-1: Uniq Violations: 16
```

In this run we ended up with 984+18 rows = 1002 rows.

```
postgres=# select count(*) from users;
 count
-------
  1002
(1 row)
```

Here is where some of the rows with duplicate ename seems to happen:
```
postgres=# postgres=# select * from users where id='user-0-27';
    id     |  ename   | age
-----------+----------+-----
 user-0-27 | name--27 |  47
(1 row)

postgres=# select * from users where id='user-1-27';
    id     |  ename   | age
-----------+----------+-----
 user-1-27 | name--27 |  47
(1 row)

postgres=# select * from users where id='user-0-27';
    id     |  ename   | age
-----------+----------+-----
 user-0-27 | name--27 |  47
(1 row)
```

The index itself seems to have duplicate entries (not just the main table).
```
postgres=# explain select count(ename) from users;
                                     QUERY PLAN
-------------------------------------------------------------------------------------
 Aggregate  (cost=6.51..6.52 rows=1 width=8)
   ->  Index Only Scan using name_idx on users  (cost=0.00..4.01 rows=1000 width=32)
(2 rows)

postgres=# select count(ename) from users;
 count
-------
  1002
(1 row)
```

Here's the table/index describe output:
```
postgres=# \d users;
               Table "public.users"
 Column |  Type   | Collation | Nullable | Default
--------+---------+-----------+----------+---------
 id     | text    |           | not null |
 ename  | text    |           |          |
 age    | integer |           |          |
Indexes:
    "users_pkey" PRIMARY KEY, lsm (id)
    "name_idx" UNIQUE, lsm (ename)

postgres=# \d name_idx;
      Index "public.name_idx"
 Column | Type | Key? | Definition
--------+------+------+------------
 ename  | text | yes  | ename
unique, lsm, for table "public.users"

```

**Comment by kmuthukk** [comment]
Aside from the data inconsistency we are creating here, we should also not be surfacing this error:

`Unexpected exception: Query error: Restart read required at: { read: { physical: 1558746958970781 } local_limit: { physical: 1558746958971794 } global_limit: <min> in_txn_limit: <max> serial_no: 0 }`

and the restart for this case should be internal to the system.

**Comment by d-uspenskiy** [comment]
Issue was fixed in context of #1058 
Commit is https://github.com/YugaByte/yugabyte-db/commit/abb09b1001e1875abce6c1c491c4140eda8e9bf7
